roc.go

```diff
@@ -0,0 +1,135 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package stat
+
+import "sort"
+
+// ROC returns paired false positive rate (FPR) and true
+// positive rate (TPR) values corresponding to n
+// cutoffs spanning the relative (or receiver)
+// operator characteristic (ROC) curve obtained when y is
+// treated as a binary classifier for c with weights w.
+// Cutoffs are equally spaced from eps less than the minimum value of y
+// to the maximum value of y, including both endpoints -- this means the
+// resulting ROC curve will always begin at (0,0) and end at (1,1).
+// input y must be sorted, and SortWeightedLabelled can be used
+// in order to sort data y together with boolean labels c and weights w.
+// For a given cutoff value, observations corresponding to
+// entries in y greater than the cutoff are classified as
+// false, and these assigned class labels are compared to the
+// class labels in c to calculate FPR and TPR.
+// w being nil results in the unweighted case.
+// n of zero results in all possible cutoffs being calculated.
+// More details about ROC curves are available at
+// https://en.wikipedia.org/wiki/Receiver_operating_characteristic
+func ROC(y []float64, c []bool, w []float64, n int) (tpr, fpr []float64) {
+	if len(y) != len(c) {
+		panic("stat: slice length mismatch")
+	}
+	if !sort.IsSorted(sort.Float64Slice(y)) {
+		panic("stat: input must be sorted")
+	}
+
+	var j int
+	var np, nn, iw, tol float64
+	if n != 0 {
+		if n < 2 {
+			panic("stat: cannot less than 2 points on a ROC curve")
+		}
+		iw = (y[len(y)-1] - y[0]) / float64(n-1)
+		if iw <= 0 {
+			panic("stat: y constant or not sorted")
+		}
+		tol = y[0] + iw
+	}
+
+	ne := true
```

Comment:
If it means notEqual, rename it to equal and invert the logic. Negations are a bad idea most of the time.